### PR TITLE
Refactor index helpers

### PR DESCRIPTION
**Pull request by jaimergp**
This PR adds a fallback behaviour for conda channels the libmamba repodata fetchers don't support yet (e.g. `s3://`). It would be trivial to adapt this other rules if needed (e.g. on-prem stuff, special auth).

If the channel contains a scheme and this scheme is not http[s] or file, we will delegate to conda's `SubdirData`. To ensure priorities are respected, we fetch channels in order. 

Side note: the adapter code used would allow to totally get by without libmamba loaders if needed, too.

Note conda doesn't have tests where s3 channels are actually fetched.

### CODE REVIEW

File: tests/test_channels.py
Comment (jaimergp) [inline tests/test_channels.py:144]: @jezdez I am wondering whether this mocked channel tests would be useful upstream too. Most tests just use anaconda.org and that's it. We don't have S3 or FTP specific tests in `conda/conda`.

Comment (jezdez) [reply tests/test_channels.py:144]: Yes, that'd be splendid!

**Comment by jaimergp** [comment]
I am [upstreaming the S3 tests to conda/conda](https://github.com/conda/conda/pull/11644). Once that one is merged, we can remove it here, so let's not merge this one yet!

**Comment by jaimergp** [comment]
Hm, unclear what's wrong now, since the test is essentially the same?

For some reason the download is expected to be empty, but then it's not, but then some files are reported missing and things break 🤷 

I'll take a look tomorrow, with a fresher mind.

File: .github/workflows/upstream_ci.yml
Comment (jezdez) [inline .github/workflows/upstream_ci.yml:130]: This makes it fail if there is no `dev/macos/setup.sh`?

File: tests/channel_testing_utils.py
Comment (jezdez) [inline tests/channel_testing_utils.py:14]: Just tiny naming thing, since this is not related to the `mock` library I'd rather not use it for this fixture to reduce confusion when we use it in the future.

**Review by jezdez — APPROVED** [review]
Looks good to me, just minor questions.

File: .github/workflows/upstream_ci.yml
Comment (jaimergp) [reply .github/workflows/upstream_ci.yml:130]: Yes. This is an oversight we had at `setup-miniconda` and I am slowly adopting the new `shell` across projects. [See issue here](https://github.com/conda-incubator/setup-miniconda/issues/227).

**Pull request merged**
